## Assignment 2 - Segmenting and Clustering Neighborhoods in Toronto - Part 2 

Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.

In an older version of this course, we were leveraging the Google Maps Geocoding API to get the latitude and the longitude coordinates of each neighborhood. However, recently Google started charging for their API: http://geoawesomeness.com/developers-up-in-arms-over-google-maps-api-insane-price-hike/, so we will use the Geocoder Python package instead: https://geocoder.readthedocs.io/index.html.

The problem with this Package is you have to be persistent sometimes in order to get the geographical coordinates of a given postal code. So you can make a call to get the latitude and longitude coordinates of a given postal code and the result would be None, and then make the call again and you would get the coordinates. So, in order to make sure that you get the coordinates for all of our neighborhoods, you can run a while loop for each postal code. Taking postal code M5G as an example, your code would look something like this:

Given that this package can be very unreliable, in case you are not able to get the geographical coordinates of the neighborhoods using the Geocoder package, here is a link to a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data

Use the Geocoder package or the csv file to create the following dataframe:

In [1]:
'''import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation

!conda install -c conda-forge geopy --yes 
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize

!conda install -c conda-forge folium=0.5.0 --yes
import folium # plotting library

print('Folium installed')
print('Libraries imported.')'''

"import requests # library to handle requests\nimport pandas as pd # library for data analsysis\nimport numpy as np # library to handle data in a vectorized manner\nimport random # library for random number generation\n\n!conda install -c conda-forge geopy --yes \nfrom geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values\n\n# libraries for displaying images\nfrom IPython.display import Image \nfrom IPython.core.display import HTML \n    \n# tranforming json file into a pandas dataframe library\nfrom pandas.io.json import json_normalize\n\n!conda install -c conda-forge folium=0.5.0 --yes\nimport folium # plotting library\n\nprint('Folium installed')\nprint('Libraries imported.')"

In [2]:
'''import geocoder # import geocoder

# initialize your variable to None
lat_lng_coords = None

# loop until you get the coordinates
while(lat_lng_coords is None):
  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
  lat_lng_coords = g.latlng

latitude = lat_lng_coords[0]
longitude = lat_lng_coords[1]'''

"import geocoder # import geocoder\n\n# initialize your variable to None\nlat_lng_coords = None\n\n# loop until you get the coordinates\nwhile(lat_lng_coords is None):\n  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))\n  lat_lng_coords = g.latlng\n\nlatitude = lat_lng_coords[0]\nlongitude = lat_lng_coords[1]"

#### Couldn't figure out the error, So I used the csv given in the link

In [3]:
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
from IPython.display import HTML
import base64, io

link = "http://cocl.us/Geospatial_data"
df1 = pd.read_csv(link)

df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [4]:
df1.shape

(103, 3)

Both the data frames have 103 rows and 3 columns

#### Read in the CSV file saved in the previous assignment 

In [5]:
link = "https://raw.githubusercontent.com/seshaphani17/Coursera_Capstone/main/Postalcode.csv"
df = pd.read_csv(link)
df.head()

Postal Code       Borough                           Neighbourhood
0        M1B   Scarborough                           Malvern, Rouge
1        M1C   Scarborough   Rouge Hill, Port Union, Highland Creek
2        M1E   Scarborough        Guildwood, Morningside, West Hill
3        M1G   Scarborough                                   Woburn
4        M1H   Scarborough                                Cedarbrae

#### Merging the two data frames together based on their Postcode 

In [6]:
df1['Latitude'] = df1['Latitude'].astype('object')
df1['Longitude'] = df1['Longitude'].astype('object')

In [7]:
df1.dtypes

Postal Code    object
Latitude       object
Longitude      object
dtype: object

In [8]:
df_new = pd.concat([df, df1],axis=1)
df_new.columns = ['Postal Code','Borough','Neighbourhood', 'Postcode', 'Latitude', 'Longitude']
df_location = df_new.drop(columns=['Postcode'])
df_location.head(10)

Postal Code       Borough                                    Neighbourhood  \
0        M1B   Scarborough                                    Malvern, Rouge   
1        M1C   Scarborough            Rouge Hill, Port Union, Highland Creek   
2        M1E   Scarborough                 Guildwood, Morningside, West Hill   
3        M1G   Scarborough                                            Woburn   
4        M1H   Scarborough                                         Cedarbrae   
5        M1J   Scarborough                               Scarborough Village   
6        M1K   Scarborough       Kennedy Park, Ionview, East Birchmount Park   
7        M1L   Scarborough                   Golden Mile, Clairlea, Oakridge   
8        M1M   Scarborough   Cliffside, Cliffcrest, Scarborough Village West   
9        M1N   Scarborough                       Birch Cliff, Cliffside West   

  Latitude Longitude  
0  43.8067  -79.1944  
1  43.7845  -79.1605  
2  43.7636  -79.1887  
3   43.771  -79.2169  
4  43.7731  -79.2395  
5  43.7447  -79.2395  
6  43.7279   -79.262  
7  43.7111  -79.2846  
8  43.7163  -79.2395  
9  43.6927  -79.2648

#### download data into csv file 

In [9]:
def create_download_link_csv(df, title = "Download CSV file", filename = "data.csv"):  
    # generate in-memory CSV, then base64-encode it
    csv = df.to_csv(index=False)
    b64 = base64.b64encode(csv.encode())
    payload = b64.decode()
    html = '<a download="{filename}" href="data:text/csv;base64,{payload}" target="_blank">{title}</a>'
    html = html.format(payload=payload,title=title,filename=filename)
    return HTML(html)

In [10]:
create_download_link_csv(df,"Download csv file","Locationcoordinates.csv")

#### Finally, check to make sure the coordinates are added as required by the question

In [11]:
# create a new test dataframe
column_names = ["Postal Code", "Borough", "Neighbourhood", "Latitude", "Longitude"]
test_df = pd.DataFrame(columns=column_names)

test_list = ["M5G", "M2H", "M4B", "M1J", "M4G", "M4M", "M1R", "M9V", "M9L", "M5V", "M1B", "M5A"]

for postcode in test_list:
    test_df = test_df.append(df_location[df_location["Postal Code"].str.contains(postcode)], ignore_index=True)
    
test_df

Postal Code            Borough  \
0         M5G   Downtown Toronto    
1         M2H         North York    
2         M4B          East York    
3         M1J        Scarborough    
4         M4G          East York    
5         M4M       East Toronto    
6         M1R        Scarborough    
7         M9V          Etobicoke    
8         M9L         North York    
9         M5V   Downtown Toronto    
10        M1B        Scarborough    
11        M5A   Downtown Toronto    

                                        Neighbourhood Latitude Longitude  
0                                  Central Bay Street   43.658  -79.3874  
1                                   Hillcrest Village  43.8038  -79.3635  
2                     Parkview Hill, Woodbine Gardens  43.7064  -79.3099  
3                                 Scarborough Village  43.7447  -79.2395  
4                                             Leaside  43.7091  -79.3635  
5                                     Studio District  43.6595  -79.3409  
6                                   Wexford, Maryvale  43.7501  -79.2958  
7   South Steeles, Silverstone, Humbergate, Jamest...  43.7394  -79.5884  
8                                       Humber Summit  43.7563   -79.566  
9   CN Tower, King and Spadina, Railway Lands, Har...  43.6289  -79.3944  
10                                     Malvern, Rouge  43.8067  -79.1944  
11                          Regent Park, Harbourfront  43.6543  -79.3606